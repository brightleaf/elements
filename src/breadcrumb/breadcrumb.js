import React from 'react'
import classnames from 'classnames'
import { Sizes } from '../modifiers'

export const BreadcrumbItem = ({ children, isActive }) => {
  const classes = {
    'is-active': isActive,
  }
  const aria = {}
  if (isActive) {
    aria['aria-current'] = 'page'
  }
  return (
    <li className={classnames(classes)} {...aria}>
      {children}
    </li>
  )
}
export const Breadcrumb = ({
  children,
  isCentered,
  isRight,
  hasArrowSeparator,
  hasBulletSeparator,
  hasDotSeparator,
  hasSucceedsSeparator,
  isSmall,
  isMedium,
  isLarge,
}) => {
  const classes = {
    'is-centered': isCentered,
    'is-right': isRight,
    'has-arrow-separator': hasArrowSeparator,
    'has-bullet-separator': hasBulletSeparator,
    'has-dot-separator': hasDotSeparator,
    'has-succeeds-separator': hasSucceedsSeparator,
    ...Sizes({ isSmall, isMedium, isLarge }),
  }
  const last = children.length - 1
  const kids = React.Children.map(children, (child, i) => {
    if (last === i) {
      return React.cloneElement(child, {
        isActive: true,
      })
    }
    return child
  })

  return (
    <nav className={classnames('breadcrumb', classes)} aria-label="breadcrumbs">
      <ul>{kids}</ul>
    </nav>
  )
}
