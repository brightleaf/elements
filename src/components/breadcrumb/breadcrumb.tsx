import React, { FC, ReactNode } from "react";
import classnames from "classnames";
import { Base, BaseProps } from "../base";

export interface BreadcrumbItemProps {
  children?: React.ReactNode;
  className?: string;
  isActive?: boolean;
}
export const BreadcrumbItem: FC<BreadcrumbItemProps> = ({
  children,
  isActive,
  className,
}) => {
  const classes = {
    "is-active": isActive,
  };

  return (
    <li
      className={classnames(className, classes)}
      aria-current={isActive ? "page" : false}
    >
      {children}
    </li>
  );
};

export interface BreadcrumbProps {
  hasArrowSeparator?: boolean;
  hasBulletSeparator?: boolean;
  hasDotSeparator?: boolean;
  hasSucceedsSeparator?: boolean;
}

export const Breadcrumb: FC<BreadcrumbProps & BaseProps> = ({
  children,
  className,
  hasArrowSeparator,
  hasBulletSeparator,
  hasDotSeparator,
  hasSucceedsSeparator,
  ...props
}) => {
  const classes = {
    "has-arrow-separator": hasArrowSeparator,
    "has-bullet-separator": hasBulletSeparator,
    "has-dot-separator": hasDotSeparator,
    "has-succeeds-separator": hasSucceedsSeparator,
  };
  const chappies = children as ReactNode[];
  const last = chappies.length - 1;
  const kids = React.Children.toArray(children).map((child, i) =>
    React.isValidElement(child)
      ? React.cloneElement(child, {
          isActive: last === i,
        })
      : child
  );

  return (
    <Base
      as="nav"
      className={classnames("breadcrumb", className, classes)}
      aria-label="breadcrumbs"
      {...props}
    >
      <ul>{kids}</ul>
    </Base>
  );
};
