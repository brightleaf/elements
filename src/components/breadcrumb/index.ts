export * from './breadcrumb.stories';
export * from './breadcrumb';
