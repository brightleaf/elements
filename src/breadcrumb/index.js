import { Breadcrumb, BreadcrumbItem } from './breadcrumb'

export { Breadcrumb, BreadcrumbItem }

export default { Breadcrumb, BreadcrumbItem }
