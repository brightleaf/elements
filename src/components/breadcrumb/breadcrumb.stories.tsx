import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { Breadcrumb, BreadcrumbItem } from './breadcrumb'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Components/Breadcrumb',
  component: Breadcrumb,
} as ComponentMeta<typeof Breadcrumb>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof Breadcrumb> = (args) => (
  <Breadcrumb
    hasArrowSeparator={args.hasArrowSeparator}
    hasBulletSeparator={args.hasBulletSeparator}
    hasDotSeparator={args.hasDotSeparator}
    hasSucceedsSeparator={args.hasSucceedsSeparator}
  >
    <BreadcrumbItem>
      <a>Home</a>
    </BreadcrumbItem>
    <BreadcrumbItem isActive>
      <a>Components</a>
    </BreadcrumbItem>
    <BreadcrumbItem isActive>
      <a>Hero</a>
    </BreadcrumbItem>
  </Breadcrumb>
)

export const SeperatorOptions = Template.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
SeperatorOptions.args = {
    hasArrowSeparator: false,
    hasBulletSeparator: false,
    hasDotSeparator: false,
    hasSucceedsSeparator: false,
}
