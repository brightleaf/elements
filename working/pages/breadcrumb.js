/* eslint-disable spellcheck/spell-checker */
import React from 'react'
import {
  Breadcrumb,
  BreadcrumbItem,
  Column,
  Columns,
  Content,
  FullColumn,
  Hero,
  HeroBody,
  Title,
  SubTitle,
  Section,
} from '../../src'
import { ComponentsTabs } from '../components/tabs'

import { Snippet } from '../components/snippet'
import Highlight from '../components/highlighter'

export default () => {
  return (
    <Section>
      <Hero isLight isBold>
        <HeroBody>
          <Title>Breadcrumb</Title>
          <SubTitle as="p" is="4">
            A simple <strong>breadcrumb</strong> component to improve your
            navigation experience
          </SubTitle>
        </HeroBody>
      </Hero>

      <ComponentsTabs />
      <FullColumn>
        <Content>
          <p>
            The <strong>Breadcrumb</strong> component only requires a{' '}
            <code>Breadcrumb</code> container and a <code>BreadcrumbItem</code>{' '}
            list.
          </p>
          <p>
            The dividers are automatically created in the content of the{' '}
            <code>::before</code> pseudo-element of <code>BreadcrumbItem</code>{' '}
            tags.
          </p>
          <p>
            You can inform the current page using the <code>isActive</code>{' '}
            modifier in a <code>BreadcrumbItem</code> tag. It will disable the
            navigation of inner links.
          </p>
        </Content>
      </FullColumn>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <FullColumn>
        <Title as="h3" is="4">
          Alignment
        </Title>
        <Content>
          <p>
            For alternative alignments, use the <code>isCentered</code> and{' '}
            <code>isRight</code> attributes on the <code>Breadcrumb</code>{' '}
            component.
          </p>
        </Content>
      </FullColumn>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb isRight>
              <BreadcrumbItem>
                <a>Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb isRight>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb isCentered>
              <BreadcrumbItem>
                <a>Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb isCentered>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <FullColumn>
        <Title as="h3" is="4">
          Alternative separators
        </Title>
        <Content>
          <p>
            You can choose between <strong>4 additional separators</strong>:{' '}
            <code>hasArrowSeparator</code> <code>hasBulletSeparator</code>{' '}
            <code>hasDotSeparator</code> and <code>hasSucceedsSeparator</code>.
          </p>
        </Content>
      </FullColumn>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb hasArrowSeparator>
              <BreadcrumbItem>
                <a>Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb hasArrowSeparator>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb hasBulletSeparator>
              <BreadcrumbItem>
                <a>Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb hasBulletSeparator>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb hasDotSeparator>
              <BreadcrumbItem>
                <a>Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb hasDotSeparator>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb hasSucceedsSeparator>
              <BreadcrumbItem>
                <a>Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a>Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb hasSucceedsSeparator>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <FullColumn>
        <Title as="h3" is="4">
          Sizes
        </Title>
        <Content>
          <p>
            You can choose between <strong>3 additional sizes</strong>:{' '}
            <code>isSmall</code> <code>isMedium</code> and <code>isLarge</code>.
          </p>
        </Content>
      </FullColumn>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb isSmall>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb isSmall>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb isMedium>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb isMedium>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
      <Snippet>
        <Columns>
          <Column isOneThird>
            <Breadcrumb isLarge>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Home</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Components</a>
              </BreadcrumbItem>
              <BreadcrumbItem>
                <a href="#/breadcrumb">Hero</a>
              </BreadcrumbItem>
            </Breadcrumb>
          </Column>
          <Column isTwoThirds>
            <Highlight className="javascript" languages={['javascript']}>
              {`
import React from 'react'
import { Breadcrumb, BreadcrumbItem } from '@brightleaf/elements'

export default () => {
  return (
    <Breadcrumb isLarge>
      <BreadcrumbItem>
        <a>Home</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Components</a>
      </BreadcrumbItem>
      <BreadcrumbItem>
        <a>Hero</a>
      </BreadcrumbItem>
    </Breadcrumb>
  )
}
`}
            </Highlight>
          </Column>
        </Columns>
      </Snippet>
    </Section>
  )
}
