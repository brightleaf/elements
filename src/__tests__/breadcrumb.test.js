import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { toHaveClass } from '@testing-library/jest-dom'
import { Breadcrumb, BreadcrumbItem } from '../breadcrumb'

expect.extend({ toHaveClass })

describe('Breadcrumb component', () => {
  afterEach(cleanup)
  it('should render', () => {
    const { container } = render(
      <Breadcrumb>
        <BreadcrumbItem>
          <a>Home</a>
        </BreadcrumbItem>
        <BreadcrumbItem>
          <a>Components</a>
        </BreadcrumbItem>
        <BreadcrumbItem>
          <a>Hero</a>
        </BreadcrumbItem>
      </Breadcrumb>
    )
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('breadcrumb')
  })
})
