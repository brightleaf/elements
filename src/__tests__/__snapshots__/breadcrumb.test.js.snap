// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb component should render 1`] = `
<nav
  aria-label="breadcrumbs"
  class="breadcrumb"
>
  <ul>
    <li
      class=""
    >
      <a>
        Home
      </a>
    </li>
    <li
      class=""
    >
      <a>
        Components
      </a>
    </li>
    <li
      aria-current="page"
      class="is-active"
    >
      <a>
        Hero
      </a>
    </li>
  </ul>
</nav>
`;
